ocaml/xapi/helpers.ml

```diff
@@ -1180,24 +1180,18 @@ let retry ~__context ~doc ?(policy = Policy.standard) f =
   TaskHelper.set_cancellable ~__context;
 
   let rec loop state =
-    let result = ref None in
-    let state = ref state in
-    while !result = None do
-      try
-        if TaskHelper.is_cancelling ~__context then begin
-          error "%s locking failed: task has been cancelled" doc;
-          TaskHelper.cancel ~__context;
-          raise (Api_errors.Server_error(Api_errors.task_cancelled, [ Ref.string_of (Context.get_task_id __context) ]))
-        end;
-        result := Some (f ())
-      with
-      | Api_errors.Server_error(code, objref :: _ ) as e when code = Api_errors.other_operation_in_progress ->
+    try
+      if TaskHelper.is_cancelling ~__context then begin
+        error "%s locking failed: task has been cancelled" doc;
+        TaskHelper.cancel ~__context;
+        raise (Api_errors.Server_error(Api_errors.task_cancelled, [ Ref.string_of (Context.get_task_id __context) ]))
+      end;
+      f ()
+    with
+    | Api_errors.(Server_error(other_operation_in_progress, objref :: _ )) as e ->
```

Comment:
I don't think that this is correct: this will match on any `Server_error` and bind the value called `other_operation_in_progress` to the error code.
Previously it was also checking that the code was matching `other_operation_in_progress`. Unfortunately without a `when` I don't know of a way in OCaml to treat this as a literal instead of a value to bind.

Reply:
Just tested, the explanation is correct.